gaia-toolbar.js

```diff
@@ -15,291 +14,361 @@ var GaiaDialogMenu = require('gaia-dialog-menu');
-  this.dialog = new GaiaDialogMenu();
-
-  this.hiddenChildren.forEach(function(el) {
-    this.dialog.appendChild(el);
-    el.classList.remove('overflowing');
-  }, this);
-
-  this.appendChild(this.dialog);
-  this.dialog.addEventListener('click', this.dialog.close.bind(this.dialog));
-  this.dialog.addEventListener('closed', this.onDialogClosed.bind(this));
-  this.dialog.open(e);
-};
-
-proto.onDialogClosed = function() {
-  this.dialog.remove();
-  this.dialog = null;
-
-  this.hiddenChildren.forEach(function(el) {
-    el.classList.add('overflowing');
-    this.insertBefore(el, this.lastChild);
-  }, this);
-
-};
-
-var template = `
-<style>
-
-/** Reset
- ---------------------------------------------------------*/
-
-::-moz-focus-inner { border: 0; }
-
-/** Host
- ---------------------------------------------------------*/
-
-.-host {
-  display: block;
-}
+  makeAccessible: function() {
+    this.setAttribute('role', 'toolbar');
 
-/** Inner
- ---------------------------------------------------------*/
+    // If there's more than one toolbar present, toolbar must have a label.
+    if (this.needsLabel()) {
+        console.error('Component not fully accessible: please provide a label' +
+          'for the toolbar itself.', this);
+    }
 
-.inner {
-  display: flex;
-  height: 45px;
-  overflow: hidden;
+    // Add expandable semantics to the … button: expanded when dialog is shown
+    // and collapsed when dialog is hidden.
+    this.els.moreButton.setAttribute('aria-expanded', !!this.dialog);
+  },
+
+  needsLabel: function() {
+    return document.getElementsByTagName('gaia-toolbar').length > 1 &&
+      !this.getAttribute('data-l10n-id') && !this.getAttribute('aria-label');
+  },
+
+  localize: function() {
+    if (this.l10nMoreId) {
+      this.els.moreButton.setAttribute('data-l10n-id', this.l10nMoreId);
+    } else if (this.l10nMoreValue) {
+      this.els.moreButton.setAttribute('aria-label', this.l10nMoreValue);
+    } else {
+      // Neither data-l10n-id nor aria-label is provided for the … button.
+      console.error('Component not fully localized: please provide a textual ' +
+        'alternative to "more" icon button.', this);
+    }
+  },
+
+  attached: function() {
+    this.style.visibility = 'hidden';
+    this.reflow_raf();
+  },
+
+  reflow: function() {
+    this.release();
+    var overflow = this.getOverflowData();
+    if (overflow.overflowed) { this.onOverflowed(overflow); }
+    this.style.visibility = '';
+  },
+
+  getOverflowData: function() {
+    var space = this.els.inner.clientWidth;
+    var child = this.children[0];
+    var overflowed = false;
+    var willFit = 0;
+    var total = 0;
+    var overflow;
+    var width;
+
+    while (child) {
+      width = child.clientWidth;
+      overflow = width + total - space;
+
+      if (overflow > 3) {
+        overflowed = true;
+        break;
+      }
+
+      willFit++;
+      total += width;
+      child = child.nextElementSibling;
+    }
```

Comment:
Unrelated: Some of this overflow logic is quite expensive and most use cases don't require it. I'm wondering if it should be opt-in only?

```html
<gaia-toolbar overflow-menu>
```